std/typecons.d

```diff
@@ -5626,17 +5626,17 @@ to deallocate the corresponding resource.
         .destroy(_refCounted._store._payload);
         static if (hasIndirections!T)
         {
-            pureGcRemoveRange(&_refCounted._store._payload);
+            (() @trusted => pureGcRemoveRange(&_refCounted._store._payload))();
         }
 
-        pureFree(_refCounted._store);
+        (() @trusted => pureFree(_refCounted._store))();
```

Comment:
`@trusted` for pureFree. The pointer is known not to be null due to isInitialized. Is this a bit that could cause problems if an "alias" to the stored object is captured elsewhere as in http://forum.dlang.org/post/n0nnu0$1tth$1@digitalmars.com?

Reply:
Yeah, this is where it breaks down. Because you can return a raw reference to the payload, you could potentially have `RefCounted` free the memory from underneath you.

Reply:
If the change were made in Array instead, the RefCounted payload is private to Array so a reference to the payload itself couldn't escape, but we're still left with the problem of an escaping reference to one of the items contained in the Array whose space will be freed when the reference count hits zero. Is there a way forward that doesn't rely on [DIP-1000](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1000.md)?

Reply:
One approach that comes to mind is making Array's `opIndex` return by value instead of by reference. To avoid breaking existing code or giving unpleasant surprises we'd have to limit this to types that are reasonably small and don't have a complex destructor and don't have a complex copy constructor. (Of course for those types there would need to be an `opIndexAssign`.) Then we know that no references to the to-be-freed memory could have escaped without unsafe shenanigans.

Reply:
If there was a way to say "in safe mode, only return values, not references", that would be ideal. I'm unaware if there is such a thing.

Reply:
One thought that occurs to me is to use `std.traits.isSafe` with a suitably-crafted lambda / function in a template sig constraint, and thereby alter the function's refness depending on where it's called from.  Unfortunately, since we can't overload on `@safe`/`@system`, I'm not sure if this is achievable.

Reply:
What I am thinking of now is adding a `Flag!"returnByValue"` parameter to the Array template that defaults to `!hasElaborateDestructor!T && !hasElaborateCopyConstructor!T && T.sizeof <= (uint[]).sizeof` but can be manually specified if desired.